lib/pure/os.nim

```diff
@@ -3217,6 +3184,52 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
         raiseOSError(osLastError(), path)
     rawToFormalFileInfo(rawInfo, path, result)
 
+proc sameFileContent*(path1, path2: string; checkSize = false; bufferSize = 8192.Positive): bool {.
+    rtl, extern: "nos$1", tags: [ReadIOEffect], noWeirdTarget.} =
+  ## Returns `true` if both pathname arguments refer to files with identical binary content.
+  ##
+  ## If `checkSize` is `true` then checks the file sizes *before reading the files*,
+  ## if the files have different file sizes they can not have the same contents,
+  ## `checkSize = true` may be faster, specially for very big files.
+  ##
+  ## .. code-block:: nim
+  ##   echo sameFileContent("file0.txt", "file1.txt", checkSize = true, bufferSize = 999)
+  ##
+  ## See also:
+  ## * `sameFile proc <#sameFile,string,string>`_
+  var a, b: File
+  var mustRead = true
+  var bufA = alloc(bufferSize)
+  var bufB = alloc(bufferSize)
+  try:  # readBuffer or open may or may not raise IOError.
+    if not open(a, path1):
+      raise newException(IOError, "Can not open file: $1" % [path1])
+      #mustRead = false
+    if not open(b, path2):
+      raise newException(IOError, "Can not open file: $1" % [path2])
+      #mustRead = false
+    if checkSize and getFileInfo(a).size != getFileInfo(b).size: mustRead = false
+    if mustRead:
+      while true:
+        var readA = readBuffer(a, bufA, bufferSize)
+        var readB = readBuffer(b, bufB, bufferSize)
+        if readA != readB: break
```

Comment:
1 final complication is short-reads, where you'd get `readA != readB` yet the files have same content, eg because of EINTR that occured for 1 of the `readBuffer`.

There are multiple correct ways to handle this, eg using: `let readN = min(readA, readB)` but it requires a few more changes in the code to be correct.

Note that this is a pre-existing issue so could be left as TODO for another PR (and this comment un-resolved)